File: LeetCode/Medium/CPP/150_Evaluate_Reverse_Polish_Notation.cpp
#include<stack>
#include<string>
#include<iostream>
#include<vector>

using namespace std;

int evalRPN(vector<string>& tokens) {
        stack<int> res;
        int size = tokens.size();
        int val1,val2;
        for (int i = 0; i < size; i++)
        {
            if (!tokens[i].compare("+"))
            {
                val1 = res.top();
                res.pop();
                val2 = res.top();
                res.pop();
                res.push(val2 + val1);
            }
            else if (!tokens[i].compare("-"))
            {
                val1 = res.top();
                res.pop();
                val2 = res.top();
                res.pop();
                res.push(val2 - val1);
            }
            else if (!tokens[i].compare("*"))
            {
                val1 = res.top();
                res.pop();
                val2 = res.top();
                res.pop();
                res.push(val2 * val1);
            }
            else if (!tokens[i].compare("/"))
            {
                val1 = res.top();
                res.pop();
                val2 = res.top();
                res.pop();
                res.push(val2 / val1);
            }
            else
            {
                res.push(stoi(tokens[i]));
            }
        }
        return (res.top());
    }

int main(void)
{
    vector<string> tokens = {"2", "1", "+", "3", "*"};
    cout << evalRPN(tokens);
    return (0);
}

File: LeetCode/Medium/PYTHON/150_Evaluate_Reverse_Polish_Notation.py
## Time Complexity : O(n)

class Solution:
    def evalRPN(self, tokens) -> int:
        stack = []
        for c in tokens:
            if c == "+":
                stack.append(stack.pop() + stack.pop())
            elif c == "-":
                a, b = stack.pop(), stack.pop()
                stack.append(b - a)
            elif c == "*":
                stack.append(stack.pop() * stack.pop())
            elif c == "/":
                a, b = stack.pop(), stack.pop()
                stack.append(int(b / a))
            else:
                stack.append(int(c))
        return stack[0]


a = Solution()

print(a.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))